File: book/swidgets/c#/SWidgets/STextBox.cs
using System;
using System.Collections.Generic;
using System.Windows.Controls;
using Sodium.Frp;

namespace SWidgets
{
    public class STextBox : TextBox, IDisposable
    {
        private readonly Action disposeListeners;

        public STextBox(string initText)
            : this(Stream.Never<string>(), initText)
        {
        }

        public STextBox(string initText, Cell<bool> enabled)
            : this(Stream.Never<string>(), initText, enabled)
        {
        }

        public STextBox(Stream<string> setText, string initText)
            : this(setText, initText, Cell.Constant(true))
        {
        }

        public STextBox(Stream<string> setText, string initText, Cell<bool> enabled)
        {
            base.Text = initText;

            List<IListener> listeners = new List<IListener>();

            StreamSink<int> sDecrement = Stream.CreateSink<int>();
            Cell<bool> allow = setText.Map(_ => 1).OrElse(sDecrement).Accum(0, (b, d) => b + d).Map(b => b == 0);

            StreamSink<string> sUserChanges = Stream.CreateSink<string>();
            this.SUserChanges = sUserChanges;
            this.Text = sUserChanges.Gate(allow).OrElse(setText).Hold(initText);

            void TextChangedEventHandler(object sender, TextChangedEventArgs args)
            {
                string text = base.Text;
                this.Dispatcher.InvokeAsync(() => sUserChanges.Send(text));
            }

            this.TextChanged += TextChangedEventHandler;

            // Set the initial value at the end of the transaction so it works with CellLoops.
            Transaction.Post(() => this.Dispatcher.InvokeIfNecessary(() => this.IsEnabled = enabled.Sample()));

            listeners.Add(setText.Listen(t =>
            {
                this.Dispatcher.InvokeAsync(() =>
                {
                    this.TextChanged -= TextChangedEventHandler;
                    base.Text = t;
                    this.TextChanged += TextChangedEventHandler;
                    sDecrement.Send(-1);
                });
            }));

            listeners.Add(enabled.Updates().Listen(e => this.Dispatcher.InvokeIfNecessary(() => this.IsEnabled = e)));

            this.disposeListeners = () =>
            {
                foreach (IListener l in listeners)
                {
                    l.Unlisten();
                }
            };
        }

        public void Dispose()
        {
            this.disposeListeners();
        }

        public new Cell<string> Text { get; }
        public Stream<string> SUserChanges { get; }
    }
}

File: book/swidgets/f#/SWidgets/STextBox.fs
namespace SWidgets

open System
open System.Windows.Controls
open Sodium.Frp
open SWidgets.DispatcherExtensionMethods

type STextBox(setText : Stream<string>, initText : string, enabled : Cell<bool>) as this =
    inherit TextBox()

    let mutable textChangedEventHandler = Option<IDisposable>.None

    let init () =
        let sDecrement = sinkS ()
        let allow = (setText |> mapToS 1, sDecrement) |> orElseS |> accumS 0 (+) |> mapC ((=) 0)
        postT (fun () -> this.Dispatcher.InvokeIfNecessary (fun () -> this.IsEnabled <- enabled |> sampleC))
        let sUserChanges = sinkS ()
        let text = (sUserChanges |> gateC allow, setText) |> orElseS |> holdS initText
        let subscribe () = 
            this.TextChanged.Subscribe (fun _ ->
                let text = this.BaseText
                this.Dispatcher.InvokeAsync(fun () -> sUserChanges |> sendS text) |> ignore)
        let listener = setText |> listenS (fun t ->
            this.Dispatcher.InvokeAsync(fun () ->
                match textChangedEventHandler with | None -> () | Some h -> h.Dispose()
                this.BaseText <- t
                textChangedEventHandler <- Some (subscribe ())
                sDecrement |> sendS -1) |> ignore)
        let listener = Listener.append listener (enabled |> updatesC |> listenS (fun e -> this.Dispatcher.InvokeIfNecessary(fun () -> this.IsEnabled <- e)))
        sUserChanges, text, (fun () -> listener |> unlistenL), subscribe

    let sUserChanges, text, disposeListener, subscribe = init ()

    do
        base.Text <- initText
        textChangedEventHandler <- Some (subscribe ())

    new(setText, initText) = new STextBox(setText, initText, constantC true)
    new(initText, enabled) = new STextBox(neverS (), initText, enabled)
    new(initText) = new STextBox(neverS (), initText)

    member private __.BaseText
        with get () = base.Text
        and set value = base.Text <- value

    member val Text = text
    member val SUserChanges = sUserChanges

    interface IDisposable with
        member __.Dispose() = disposeListener ()